consensus/core/src/transactions_client.rs

```diff
@@ -20,32 +18,45 @@ const MAX_PENDING_TRANSACTIONS: usize = 2_000;
 #[allow(dead_code)]
 pub(crate) struct TransactionsConsumer {
     tx_receiver: metered_channel::Receiver<Transaction>,
-    max_fetched_per_request: usize,
+    max_fetched_bytes_per_request: u64,
+    pending_transaction: Option<Transaction>,
 }
 
 #[allow(dead_code)]
 impl TransactionsConsumer {
-    pub(crate) fn new(tx_receiver: metered_channel::Receiver<Transaction>) -> Self {
+    pub(crate) fn new(
+        tx_receiver: metered_channel::Receiver<Transaction>,
+        max_fetched_bytes_per_request: u64,
+    ) -> Self {
         Self {
             tx_receiver,
-            max_fetched_per_request: MAX_FETCHED_TRANSACTIONS,
+            max_fetched_bytes_per_request,
+            pending_transaction: None,
         }
     }
 
-    pub(crate) fn with_max_fetched_per_request(mut self, max_fetched_per_request: usize) -> Self {
-        self.max_fetched_per_request = max_fetched_per_request;
-        self
-    }
-
-    // Attempts to fetch the next transactions that have been submitted for sequence.
+    // Attempts to fetch the next transactions that have been submitted for sequence. Also a `max_fetched_bytes` parameter
+    // is given in order to ensure up to `max_fetched_bytes` bytes of transactions are retrieved.
     pub(crate) fn next(&mut self) -> Vec<Transaction> {
         let mut transactions = Vec::new();
-        while let Ok(transaction) = self.tx_receiver.try_recv() {
+        let mut total_fetched_size: usize = 0;
+
+        if let Some(transaction) = self.pending_transaction.take() {
+            // Here we assume that a transaction can always fit in `max_fetched_bytes_per_request`
+            total_fetched_size += transaction.data().len();
             transactions.push(transaction);
+        }
 
-            if transactions.len() >= self.max_fetched_per_request {
+        while let Ok(transaction) = self.tx_receiver.try_recv() {
+            total_fetched_size += transaction.data().len();
+
+            // If we went over the max size with this transaction, just cache it for the next pull.
+            if total_fetched_size as u64 > self.max_fetched_bytes_per_request {
```

Comment:
Also check total transaction count?

Reply:
hhm I see value on this check mainly to make sure that this loop does not spend too many cycles  during proposal but it might then become a bit tricky to tune. For example what would be an acceptable threshold in order to not cancel that max fetched bytes per request? Otherwise I am now start thinking if indeed the previous approach was bit better in the sense that the pending queue could not grow while we are consuming the queue.

Reply:
This looks good now.